### PR TITLE
input: change short-press behavior for longpress device

### PR DESCRIPTION
**Pull request by deveritec-jusc**
The change introduce a direct report of "short-press" after the button is pressed (much like the normal input button). This allows to only subscribe to the longpress-device and get a "press" event after the button was pressed, not after release.

In the previous implementation, the "short-press" event was raised in tandem with the "short-release" on release of the button. In order to capture "press"-events, this would necessiate to also subscribe to the input-device.

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
Hi @deveritec-rosc, the intent of the short press code is to give the application a signal to react to short press but NOT long presses, if we do this change then that would not be possible anymore, and the short press behavior would be to mimic exactly what the driving event is doing, so this would be removing a use case without adding any (plus causing a behavior change downstream which is not a great idea).

I guess your specific application wants to always capture short presses, regardless of whether they are also long ones, but you are trying to avoid subscribing to multiple devices?

If that's the case I really think you should just register the callback with no device filter and implement an early filter in the callback itself, the overhead of doing that should be negligible. Alternatively you could register two callbacks and make them trigger a common function... that's how I envisioned the callback to be used anyway.

**Comment by deveritec-rosc** [comment]
Hej @fabiobaltieri,
in my application I check for both short-events (press/release) as well as long-events (press/release) events. (Also some timed patterns, but this is something application specific, so I wouldn't dare adding stuff like that to the driver :) ).

What is confusing to me is the "short-press" behavior:
On release of the button, I get two notifications:
1. the "short-press" event, and 
2. immediately after the "short-release" event.

From an implementation perspective, it makes sense: "When do we know that a short-press was performed? If it was released before the longpress-timeout!" But from a usage perspective, "short-press" is now issued way after the button was pressed, and holds now additional information. 

This leads me to your implementation hint:
You are absolutely right, I can always subscribe to capture changes of the input device. This way (and some application code), I could capture both the "short-press", "short-release" and the longpress-events.
But then, why do I need the short-events now at all? The binding forces me to define the short-inputs, which I later ignore by implementing my own code for short-input capture.

From the above, it looks like "short-press" is actually never a valid option. Using the longpress-device, I get this information either
- too late, or
- need to listen to the input device anyway, so I do not need any short-press information.

Regarding the change, I hope the above clears up what I mean:
though it isn't technically right to send the "short-press" on button press (since we do not know what kind of press it is), putting it on the release is also not valid. If the user is aware that he get a "short-press" before a "long-press", he could work around it.

Maybe the short-input should be removed generally, and the longpress-device only allows for "longpress" events? At least to me, this would be a clean solution, since with all the stuff I have written above short-events feel out of place in the longpress-device.

Regarding the behavior change:
Updating through a lot of Zephyr iterations, I have always to fight against behavior changes, configuration renames etc. in different minor versions. That's not cool, but it's okay, since this is a living project. As long as stuff like this is mentioned in the release notes, people are made aware and could act accordingly (e.g. by just using the "short-release" event). Oh, and since removing of the short-input would break the devicetree files, people would definetely see that something has changed (and be angry...).

**Comment by fabiobaltieri** [comment]
> But then, why do I need the short-events now at all? The binding forces me to define the short-inputs, which I later ignore by implementing my own code for short-input capture.

Ok fair point, want to make `short_codes` optional?

**Comment by deveritec-rosc** [comment]
Sounds fine to me. I would set up a new PR and will close this one accordingly. I would also add an example for the longpress-device, I think this may help future users.

One more thing: How about the "press" event. Remove as well, or just better document it?

**Comment by fabiobaltieri** [comment]
> Sounds fine to me. I would set up a new PR and will close this one accordingly. I would also add an example for the longpress-device, I think this may help future users.
> 
> One more thing: How about the "press" event. Remove as well, or just better document it?

I mean, make the whole property optional (change the property, tweak the build warning, tweak the init to instantiate the list conditionally, set the pointe null if the prop is not set, change the code to not do any short anything if there's no short code present).

Don't think it needs a dedicated sample, just tweak the description to mention it.

**Comment by deveritec-rosc** [comment]
Updated the code.
Now, "short-codes" are purely optional, and can be ignored by the user if necessary.

I really like it, since the longpress-device now defaults to just handling longpress things.

File: subsys/input/input_longpress.c [hunk 1]
```diff
@@ -108,35 +107,38 @@ static int longpress_init(const struct device *dev)
...
 	static void longpress_cb_##inst(struct input_event *evt)                                   \
 	{                                                                                          \
 		longpress_cb(DEVICE_DT_INST_GET(inst), evt);                                       \
 	}                                                                                          \
 	INPUT_CALLBACK_DEFINE(DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(inst, input)),                 \
 			      longpress_cb_##inst);                                                \
 	static const uint16_t longpress_input_codes_##inst[] = DT_INST_PROP(inst, input_codes);    \
-	static const uint16_t longpress_short_codes_##inst[] = DT_INST_PROP(inst, short_codes);    \
+	__maybe_unused static const uint16_t longpress_short_codes_##inst[] =                      \
+		DT_INST_PROP_OR(inst, short_codes, {});                                            \
```
Comment (fabiobaltieri) [inline subsys/input/input_longpress.c:123]: I think you can use a `COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, short_codes)...` here to avoid defining it entirely.

File: subsys/input/input_longpress.c [hunk 2]
```diff
@@ -108,35 +107,38 @@ static int longpress_init(const struct device *dev)
 	return 0;
 }
 
+/* short_codes are now optional, if not set the assert shouldn't fail */
+#define SHORT_CODES_LEN(inst) DT_PROP_LEN_OR(inst, short_codes, DT_INST_PROP_LEN(inst, input_codes))
+
 #define INPUT_LONGPRESS_DEFINE(inst)                                                               \
-	BUILD_ASSERT(DT_INST_PROP_LEN(inst, input_codes) ==                                        \
-		     DT_INST_PROP_LEN(inst, short_codes));                                         \
```
Comment (fabiobaltieri) [inline subsys/input/input_longpress.c:113]: `(DT_INST_PROP_LEN(inst, input_codes) == DT_INST_PROP_LEN(inst, short_codes)) || !DT_INST_NODE_HAS_PROP(inst, short_codes)`? No need to drop it

File: subsys/input/input_longpress.c [hunk 3]
```diff
@@ -108,35 +107,38 @@ static int longpress_init(const struct device *dev)
...
+	__maybe_unused static const uint16_t longpress_short_codes_##inst[] =                      \
+		DT_INST_PROP_OR(inst, short_codes, {});                                            \
 	static const uint16_t longpress_long_codes_##inst[] = DT_INST_PROP(inst, long_codes);      \
 	static const struct longpress_config longpress_config_##inst = {                           \
 		.input_dev = DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(inst, input)),                  \
 		.input_codes = longpress_input_codes_##inst,                                       \
-		.short_codes = longpress_short_codes_##inst,                                       \
+		.short_codes = (sizeof(longpress_short_codes_##inst) != 0)                         \
+				       ? longpress_short_codes_##inst                              \
+				       : NULL,                                                     \
```
Comment (fabiobaltieri) [inline subsys/input/input_longpress.c:130]: same as above

File: subsys/input/input_longpress.c [hunk 4]
```diff
@@ -108,35 +107,38 @@ static int longpress_init(const struct device *dev)
...
 		.long_codes = longpress_long_codes_##inst,                                         \
 		.num_codes = DT_INST_PROP_LEN(inst, input_codes),                                  \
 		.long_delays_ms = DT_INST_PROP(inst, long_delay_ms),                               \
 	};                                                                                         \
-	static struct longpress_data_entry longpress_data_entries_##inst[DT_INST_PROP_LEN(         \
-			inst, input_codes)];                                                       \
+	static struct longpress_data_entry                                                         \
+		longpress_data_entries_##inst[DT_INST_PROP_LEN(inst, input_codes)];                \
```
Comment (fabiobaltieri) [inline subsys/input/input_longpress.c:136]: unrelated, do not change

File: tests/subsys/input/input_longpress/src/main.c [hunk 5]
```diff
@@ -9,13 +9,13 @@
...
-		DT_NODELABEL(longpress));
+static const struct device *const fake_dev = DEVICE_DT_GET(DT_NODELABEL(fake_input_device));
+static const struct device *const longpress_dev = DEVICE_DT_GET(DT_NODELABEL(longpress));
+static const struct device *const longpress_no_short_dev =
+	DEVICE_DT_GET(DT_NODELABEL(longpress_no_short));
 
-DEVICE_DT_DEFINE(DT_INST(0, vnd_input_device), NULL, NULL, NULL, NULL,
-		 PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE, NULL);
+DEVICE_DT_DEFINE(DT_INST(0, vnd_input_device), NULL, NULL, NULL, NULL, PRE_KERNEL_1,
+		 CONFIG_KERNEL_INIT_PRIORITY_DEVICE, NULL);
```
Comment (fabiobaltieri) [inline tests/subsys/input/input_longpress/src/main.c:18]: drop unrelated change, keep the same style

File: tests/subsys/input/input_longpress/src/main.c [hunk 6]
```diff
@@ -30,6 +30,20 @@ static void test_cb(struct input_event *evt)
...
+{
+	TC_PRINT("%s: %d %x %d\n", __func__, event_count_no_short, evt->code, evt->value);
+
+	event_count_no_short++;
+	memcpy(&last_events_no_short[1], &last_events_no_short[0], sizeof(struct input_event));
+	memcpy(&last_events_no_short[0], evt, sizeof(struct input_event));
+}
+INPUT_CALLBACK_DEFINE(longpress_no_short_dev, test_cb_no_short);
+
+/* ------ */
```
Comment (fabiobaltieri) [inline tests/subsys/input/input_longpress/src/main.c:45]: drop

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]

File: subsys/input/input_longpress.c (hunk 2 above)
Comment (deveritec-rosc) [reply subsys/input/input_longpress.c:113]: Somehow I dropped this by error - readded it.

File: subsys/input/input_longpress.c [hunk 7]
```diff
@@ -108,35 +107,38 @@ static int longpress_init(const struct device *dev)
 	return 0;
 }
 
+/* short_codes are now optional, if not set the assert shouldn't fail */
+#define SHORT_CODES_LEN(inst) DT_PROP_LEN_OR(inst, short_codes, DT_INST_PROP_LEN(inst, input_codes))
+
```
[the remark this answers is not in the document]
Comment (deveritec-rosc) [reply subsys/input/input_longpress.c:112]: necessary for the length check - see reintrodution

File: subsys/input/input_longpress.c (hunk 3 above)
Comment (deveritec-rosc) [reply subsys/input/input_longpress.c:130]: hart to keep the style here, since the line length would be > 100 chars and clang formatter changes it to this style.

File: subsys/input/input_longpress.c (hunk 2 above)
Comment (fabiobaltieri) [reply subsys/input/input_longpress.c]: sure, but you should still use the condition I mentioned and drop it

File: subsys/input/input_longpress.c (hunk 1 above)
Comment (fabiobaltieri) [reply subsys/input/input_longpress.c:125]: cool, now this one is also still relevant, you can do something like

```diff
        INPUT_CALLBACK_DEFINE(DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(inst, input)),                 \
                              longpress_cb_##inst);                                                \
        static const uint16_t longpress_input_codes_##inst[] = DT_INST_PROP(inst, input_codes);    \
+       IF_ENABLED(DT_INST_NODE_HAS_PROP(inst, short_codes), (                                     \
        static const uint16_t longpress_short_codes_##inst[] = DT_INST_PROP(inst, short_codes);    \
+       ));                                                                                        \
        static const uint16_t longpress_long_codes_##inst[] = DT_INST_PROP(inst, long_codes);      \
        static const struct longpress_config longpress_config_##inst = {                           \
                .input_dev = DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(inst, input)),                  \
                .input_codes = longpress_input_codes_##inst,                                       \
+               IF_ENABLED(DT_INST_NODE_HAS_PROP(inst, short_codes), (                             \
                .short_codes = longpress_short_codes_##inst,                                       \
+               ))                                                                                 \
                .long_codes = longpress_long_codes_##inst,                                         \
                .num_codes = DT_INST_PROP_LEN(inst, input_codes),                                  \
                .long_delays_ms = DT_INST_PROP(inst, long_delay_ms),                               \
```

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@deveritec-rosc sorry there's a conflict to be resolved in the PR now, could you rebase please?

**Comment by deveritec-rosc** [comment]
done

**Review by fabiobaltieri — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@gmarull @kartben could you take a look as well please?

**Review by gmarull — APPROVED** [review]
